### PR TITLE
Health analyzer update

### PR DESCRIPTION
**Pull request by Neocloudy**
## About The Pull Request
Reworks the styling of the health analyzer's limb information and adds additional CM health analyzer features: treatment advice, hackerman (green) theme when working with robots/synths, and damage types are colored accordingly.

To implement treatment advice, I've messed with the code of health analyzers slightly, as it was impossible to work with in its previous state. `patient`/`current_user` are now at a carbon human level, and the `ishuman` check that had all the functionality of the health analyzer is turned into an early return.

Treatment advice currently lists:
- Time left until permanently dead, remove patient armor, defib when close enough to revival threshold
- Use a blowtorch/cable coil to fix a synthetic, or use trauma/burn kits to fix a person
- Sleep/cryo/use a robotic cradle when you have cellular damage or integrity loss
- Administer isotonic solution when low on blood
- Embryo advice
- Infection advice
- Chemical advice- "administer bica" "administer kelo" "administer tram" etc
  - Some entries may need more work
- Advice to remove embedded objects
- Warnings not to administer bad chemicals to nanite/paracetamol users

image
image
## Why It's Good For The Game
The new style of damaged limbs is easier to read and better looking than the current style.

Treatment advice is a classic feature, apparently intentionally removed when we converted health scanners to TGUI. I think they could be very helpful for new corpsmen to know how to treat someone.

Switching the theme of a health analyzer and hiding toxin/oxyloss when working with synthetics is a really basic feature that I've wanted since CM made their health analyzers use TGUI. This makes it really easy to identify a synthetic/robot if you're focusing on making marines vertical and not paying much attention, like me. 
## Changelog
:cl:
add: Reworked the style of the Limbs Damaged section of the health analyzer.
add: Health analyzers now have a section for relevant information such as treatment advice when scanning people again.
add: Information includes death/revival information, healing information, chemical advice, etc.
add: The theme of a health analyzer will switch to green when working with a synthetic or combat robot.
qol: Robotic/Biotic limbs are shown in Damaged Limbs.
qol: Revivable status now includes information such as too much damage, etc.
qol: Toxin and Oxyloss will be hidden when working with a synthetic or combat robot.
/:cl:

### CODE REVIEW

**Comment by Neocloudy** [comment]
lmfao we don't have peridaxon anymore

**Comment by lbnesquik** [comment]
We only have peri+ I think. Also, wow! Making it easy!

**Comment by Jackal-boop** [comment]
could you make the extra info collapsable?

**Comment by Arctrooper8** [comment]
You don't mention anything about Saline Glucose, Nanoblood, or IV drips with blood bags for the health tip about blood loss.

**Comment by Neocloudy** [comment]
> You don't mention anything about Saline Glucose, Nanoblood, or IV drips with blood bags for the health tip about blood loss.

mentioning all of those would probably be very bulky, i'll change the food advice to isotonic solution
image

**Comment by Neocloudy** [comment]
> could you make the extra info collapsable?

Late response, but this is *probably* something I can do, I just don't know too much about TGUI to do it. Probably in a future PR.

File: code/game/objects/items/scanners.dm [hunk 1]
```diff
@@ -186,88 +186,292 @@ REAGENT SCANNER
...
+			"color" = "white"
+			))
+	if(advice.len)
+		data["advice"] = advice
+	else
+		data["advice"] = null
+
 	var/ssd = null
 	if(patient.has_brain() && patient.stat != DEAD && ishuman(patient))
 		if(!patient.key)
```
[the remark this answers is not in the document]
Comment (Xander3359) [reply code/game/objects/items/scanners.dm:471]: We checked ishuman() earlier

File: code/game/objects/items/scanners.dm [hunk 2]
```diff
@@ -186,88 +186,292 @@ REAGENT SCANNER
...
+				if(0.4 * TIME_BEFORE_DNR to 0.8 * TIME_BEFORE_DNR)
+					dead_color = "orange"
+				if(0.8 * TIME_BEFORE_DNR to INFINITY)
+					dead_color = "red"
+			if(!issynth(patient)) // synthetics don't expire
+				advice += list(list(
+					"advice" = "Time remaining to revive: [DisplayTimeText((TIME_BEFORE_DNR-(patient.dead_ticks))*20)].",
+					"icon" = "clock",
+					"color" = dead_color
+					))
```
Comment (Xander3359) [inline code/game/objects/items/scanners.dm:293]: What happens if I scan a synth.
Do they have dead ticks?

**Review by Xander3359 — CHANGES_REQUESTED** [review]

File: code/game/objects/items/scanners.dm [hunk 3]
```diff
@@ -186,88 +186,292 @@ REAGENT SCANNER
...
+
+	var/infection_message
+	var/internal_bleeding
+
+	var/unknown_implants = 0
+	for(var/datum/limb/limb AS in patient.limbs)
+		var/infected = FALSE
+		var/necrotized = FALSE
+
+		if(!internal_bleeding)
```
[the remark this answers is not in the document]
Comment (Lumipharon) [reply code/game/objects/items/scanners.dm:200]: this is in fact, incorrect and works fine.

File: code/game/objects/items/scanners.dm [hunk 4]
```diff
@@ -186,88 +186,292 @@ REAGENT SCANNER
...
+	else
+		advice += list(list(
+			"advice" = "Patient is unrevivable.",
+			"icon" = "ribbon",
+			"color" = "white"
+			))
+	if(advice.len)
+		data["advice"] = advice
+	else
+		data["advice"] = null
```
[the remark this answers is not in the document]
Comment (Lumipharon) [reply code/game/objects/items/scanners.dm:467]: ?? why

File: tgui/packages/tgui/interfaces/MedScanner.jsx [hunk 5]
```diff
@@ -41,11 +44,15 @@ export const MedScanner = (props) => {
     internal_bleeding,
     implants,
     hugged,
+    advice,
   } = data;
   const chemicals = Object.values(chemicals_lists);
   const limb_data = Object.values(limb_data_lists);
+  let index = 0;
```
Comment (TiviPlus) [inline tgui/packages/tgui/interfaces/MedScanner.jsx:51]: name is var better pls

File: tgui/packages/tgui/interfaces/MedScanner.jsx [hunk 6]
```diff
@@ -173,126 +159,108 @@ export const MedScanner = (props) => {
...
+                      <Flex.Item shrink="1">
+                        {limb.limb_status ? (
+                          <Box
+                            inline
+                            color={
+                              limb.limb_status === 'Splinted' ? 'lime' : 'white'
+                            }
+                            bold={1}
+                          >
+                            [{limb.limb_status}]
```
Comment (TiviPlus) [inline tgui/packages/tgui/interfaces/MedScanner.jsx:219]: Why is all this not using Stack?

**Review by TiviPlus — CHANGES_REQUESTED** [review]
Please do not self resolve reviews, especially without comment and extra especially without updating anything

File: code/game/objects/items/scanners.dm [hunk 7]
```diff
@@ -162,7 +162,7 @@ REAGENT SCANNER
 		"oxy" = round(patient.getOxyLoss()),
 		"clone" = round(patient.getCloneLoss()),
 
-		"revivable" = patient.getBruteLoss() + patient.getFireLoss() + patient.getToxLoss() + patient.getOxyLoss() + patient.getCloneLoss() <= 200,
+		"revivable" = patient.getBruteLoss() + patient.getFireLoss() + patient.getToxLoss() + patient.getCloneLoss() <= 215, // around the revive threshold for people with corpsman level skills
```
Comment (Xander3359) [inline code/game/objects/items/scanners.dm:165]: Stop hard coding this, set it to a formula that pulls from relevant data.
If medical skill ever changes, this is no longer valid
If defib healing changes, this is no longer valid.
If revivable threshold changes, this is no longer valid.

Comment (Lumipharon) [reply code/game/objects/items/scanners.dm:165]: yeah that's stinky. this is also assuming the maxhp never changes as well.

**Comment by lbnesquik** [comment]
I think it will be better with a collapse, but I am still here for it.

File: code/game/objects/items/defibrillator.dm [hunk 8]
```diff
@@ -165,15 +171,15 @@
...
+	if(medical_skill < SKILL_MEDICAL_PRACTICED)
 		user.visible_message(span_notice("[user] fumbles around figuring out how to use [src]."),
 		span_notice("You fumble around figuring out how to use [src]."))
-		var/fumbling_time = SKILL_TASK_AVERAGE - (SKILL_TASK_VERY_EASY * skill) // 3 seconds with medical skill, 5 without
+		var/fumbling_time = SKILL_TASK_AVERAGE - (SKILL_TASK_VERY_EASY * medical_skill) // 3 seconds with medical medical_skill, 5 without
 		if(!do_after(user, fumbling_time, NONE, H, BUSY_ICON_UNSKILLED))
 			return
-	else
-		defib_heal_amt *= skill * 0.5 //more healing power when used by a doctor (this means non-trained don't heal)
+	else // for ye whom may venture here: if you kill this "else", kill the special snowflake bit in check_revive()
```
Comment (Lumipharon) [inline code/game/objects/items/defibrillator.dm:181]: the fact you're putting this note here should tell you that what you're doing is wrong.
Make the define actually correct and don't gate it behind an else at all, then you don't have to worry about someone breaking health scanners in the future

File: code/modules/mob/living/carbon/human/human_helpers.dm [hunk 9]
```diff
@@ -262,3 +262,17 @@
...
+	if(HAS_TRAIT(src, TRAIT_UNDEFIBBABLE))
+		return FALSE
+	if(HAS_TRAIT(src, TRAIT_IMMEDIATE_DEFIB))
+		return TRUE
+	var/actual_healing_level //leftover snowflake from defibrillators healing 8 for untrained/novice
+	if(user_skill < SKILL_MEDICAL_PRACTICED)
+		actual_healing_level = 8
+	testing(actual_healing_level)
+	if(health + getOxyLoss() + actual_healing_level ? actual_healing_level : (DEFIBRILLATOR_HEALING_TIMES_SKILL(user_skill)) >= get_death_threshold())
+		return TRUE
```
Comment (Lumipharon) [inline code/modules/mob/living/carbon/human/human_helpers.dm]: why is this a proc at all? Its used specifically in one place and that's it, there's no particular benefit to making it a separate proc unless its actually used elsewhere.

**Review by Lumipharon — CHANGES_REQUESTED** [review]

File: code/game/objects/items/defibrillator.dm (hunk 8 above)
Comment (Neocloudy) [reply code/game/objects/items/defibrillator.dm:181]: qv said removing the else is balance last time i tried this, but sure

Comment (Lumipharon) [reply code/game/objects/items/defibrillator.dm:181]: You have to change the define, not just remove the else

File: code/game/objects/items/defibrillator.dm [hunk 10]
```diff
@@ -1,7 +1,7 @@
-///The base healing number for a defibrillator. Kept global to be used by `check_revive()`.
+///The base healing number for a defibrillator. Needs to be global for `DEFIBRILLATOR_HEALING_TIMES_SKILL` to work outside of this file.
 #define DEFIBRILLATOR_BASE_HEALING_VALUE 8
-///The formula for healing with a defibrillator. Kept global to be used by `check_revive()`.
-#define DEFIBRILLATOR_HEALING_TIMES_SKILL(skill_input) (DEFIBRILLATOR_BASE_HEALING_VALUE * skill_input * 0.5)
+///The formula for healing with a defibrillator. If `skill_input` in this macro is less than `SKILL_MEDICAL_PRACTICED`, this will be 8.
+#define DEFIBRILLATOR_HEALING_TIMES_SKILL(skill_input) (skill_input < SKILL_MEDICAL_PRACTICED ? 8 : DEFIBRILLATOR_BASE_HEALING_VALUE * skill_input * 0.5)
```
Comment (Lumipharon) [inline code/game/objects/items/defibrillator.dm]: you'll probably want to move these defines to an actual define file now its being used in different files. Defines need to actually be defined before the code they're used in, so it breaks if someone tries to use this in a proceeding file.

**Pull request merged**